### DISCUSSION TITLE
Is there a way to automatically detect the phases for each circuit?

### DISCUSSION THREAD

**Discussion by ankushg** [emporia-vue-local/esphome#213] [Q&A]
It seems that the official Emporia app doesn't require you to specify which phase each CT clamp is associated with. This lets you avoid recompiling/reflashing just to adjust for cable management purposes

Any idea how they do that? 

Is there a way to get the Emporia Vue esphome config to auto-detect phase in that way?

**Comment by flaviut** [comment]
Here's what the raw data looks like:

![image](https://github.com/emporia-vue-local/esphome/assets/5213469/2c9c7fd5-06b7-4322-b099-36e2ababd160)

https://docs.google.com/spreadsheets/d/1Yu98tDRhFJqqMntNak6X297SKyzQtf0_Cvv_g7qKewk/edit#gid=0

It's definitely possible to guess, but it's lots of coding work and still requires the user to verify everything.

**Reply by gekkehenkie11** [reply]
> It's definitely possible to guess, but it's lots of coding work and still requires the user to verify everything. 

But that can't be true. I mean the ESP code obvioulsy just get this same data and then makes the decision which of the three PV values to use, without the user verifying anything. I think this is what Emporia calls 'locking the channels', which btw also in their process actually isn't 100% error free, so thinking of that, what you're saying is that if you want to do it even BETTER than what emporia does, then yes, the user would need to verify things.

BTW, Emporia themselves suggest to turn on all equipment in your house when having the software calculate what is what. And I guess they then just take the most positive value each time?

**Comment by ankushg** [comment]
Interesting… any idea how the stock firmware/app is able to figure it out?

**Reply by gekkehenkie11** [reply]
Like said in my above reply, Emporia suggest to turn on all equipment in your house when the firmware does the 'channel locking', as how they call it, which is the process of deciding which phase belongs to which of the 16 smaller CT's. I think after that they just take the most positive value. Which, again, might not be errorfree, you can call emporia helpline and they can then change that online for you if you think the automatic process made an error.

**Comment by kahrendt** [comment]
Could it be as easy as using the most positive of the three readings for each CT clamp? Looking quickly through that spreadsheet and the reversing repo's examples, it seems like that could work. It seems usually two out of three are negative, but if two are positive, then the one that is larger looks like the desired value. There may be some issues for very small power measurements however; i.e., circuits only using 1 or 2 watts at a given point in time. Also there could be some issues between 2-phase and 3-phase power setups that I am not considering.

**Comment by henfri** [comment]
Does the device actually report power, or does it ignore phase shift?

**Reply by flaviut** [reply]
It calculates real power—see https://github.com/emporia-vue-local/esphome/discussions/213#discussioncomment-7118267 for what the raw data looks like.

**Reply by henfri** [reply]
But is that also transmitted to HA?

There are many posts that report that apparent power is included. 
I am now not sure, what to believe...

**Reply by flaviut** [reply]
Sorry, I misunderstood your question. Only apparent power is included.

See https://github.com/emporia-vue-local/esphome/discussions/2 for more details.

**Reply by henfri** [reply]
Hm, in Europe, true Power is billed, apparent power not.
So, it is pretty useless here, then, right?

This does not seem to be a limitation of the esphome firmware, if I interpret https://github.com/emporia-vue-local/esphome/discussions/2 right ...

Why does it need a connection to each phase then? And why do I need to specify what phase each clamp is on?

In the screenshot above, it seems that three power values are calculated, one assuming per (assumed) phase...

Just to be sure:
True Power (W) = Voltage* Current * cos(phase angle)
Apparent Power (VA)= Voltage* Current

**Reply by henfri** [reply]
What is the meaning of P(V1) in contrast to P(V2) and P(V3) above?

**Comment by henfri** [comment]
Hello,

well, as the App does not allow an association of phase <-> channel, there must be some "magic". Probably, that was seen too complex for the average user.
Such magic is not too difficult, if the device is drawing some power, one could pick the one with the most positive power. 
However, one would have to keep that one ("lock") it. 

Greetings,
Hendrik

**Comment by clowrey** [comment]
I had a necessity for auto phase - as it was driving me crazy - off grid system and we kept moving the breakers around to balance loads.. 

AI to the rescue Cursor IDE + Claude Sonnet 4.5 Thinking wrote this to auto detect it... Working perfectly so far ;) but of course use at your own risk.. 

```yaml
esphome:
  name: emporia-vue
  friendly_name: Emporia Vue

external_components:
  - source: github://emporia-vue-local/esphome@dev
    components:
      - emporia_vue

# Enable logging
logger:
  logs:
    # by default, every reading will be printed to the UART, which is very slow
    # This will disable printing the readings but keep other helpful messages
    sensor: ERROR  # Changed to ERROR to reduce overhead from auto-phase detection

esp32:
  board: esp32dev
  framework:
    type: esp-idf
    version: recommended
    
# Enable Home Assistant API
api:
  encryption:
    key: !secret api_encryption_key
  reboot_timeout: 0s  # Disable auto-reboot on API disconnect
    
  services:
    - service: play_rtttl
      variables:
        song_str: string
      then:
        - rtttl.play:
            rtttl: !lambda 'return song_str;'

ota:
  - platform: esphome
    password: !secret ota_password


wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password
# Enable fallback hotspot (captive portal) in case wifi connection fails
  ap:
    ssid: "Emporia Fallback Hotspot"
    password: "fallback"

captive_portal:

preferences:
  # the default of 1min is far too short--flash chip is rated
  # for approx 100k writes.
  flash_write_interval: "48h"

output:
  - platform: ledc
    pin: GPIO12
    id: buzzer
  - platform: gpio
    pin: GPIO27
    id: buzzer_gnd

rtttl:
  output: buzzer
  on_finished_playback:
    - logger.log: 'Song ended!'

button:
  - platform: template
    name: "Two Beeps"
    on_press:
      - rtttl.play: "two short:d=4,o=5,b=100:16e6,16e6"

i2c:
  sda: 21
  scl: 22
  scan: false
  frequency: 400kHz
  timeout: 1ms
  id: i2c_a

time:
  - platform: homeassistant
    id: my_time

# these are called references in YAML. They allow you to reuse
# this configuration in each sensor, while only defining it once
.defaultfilters:
  - &throttle_avg
    # average all raw readings together over a 5 second span before publishing
    throttle_average: 5s
  - &throttle_avg_1s
    # average all raw readings together over a 5 second span before publishing
    throttle_average: 1s
  - &throttle_time
    # only send the most recent measurement every 60 seconds
    throttle: 60s
  - &throttle_internal
    # throttle internal sensors to reduce load
    throttle: 1s
  - &invert
    # invert and filter out any values below 0.
    lambda: 'return max(-x, 0.0f);'
  - &pos
    # filter out any values below 0.
    lambda: 'return max(x, 0.0f);'
  - &abs
    # take the absolute value of the value
    lambda: 'return abs(x);'
  - &kW
    multiply: .001

sensor:
  - platform: emporia_vue
    i2c_id: i2c_a
    phases:
      - id: phase_a  # Verify that this specific phase/leg is connected to correct input wire color on device listed below
        input: BLACK  # Vue device wire color
        calibration: 0.022166  # 0.022 is used as the default as starting point but may need adjusted to ensure accuracy
        # To calculate new calibration value use the formula <in-use calibration value> * <accurate voltage> / <reporting voltage>
        voltage:
          name: "Voltage L1"
          id: phase_a_volt
          filters: [ *throttle_avg, *pos ]
        frequency:
          name: "Frequency L1"
          filters: [ *throttle_avg, *pos ]

      - id: phase_b  # Verify that this specific phase/leg is connected to correct input wire color on device listed below
        input: RED  # Vue device wire color
        calibration: 0.022166  # 0.022 is used as the default as starting point but may need adjusted to ensure accuracy
        voltage:
          name: "Voltage L2"
          id: phase_b_volt
          filters: [ *throttle_avg, *pos ]
        phase_angle:
          name: "Phase Angle L2"
          filters: [ *throttle_avg, *pos ]

      - id: phase_c  # Verify that this specific phase/leg is connected to correct input wire color on device listed below
        input: BLUE  # Vue device wire color
        calibration: 0.022166  # 0.022 is used as the default as starting point but may need adjusted to ensure accuracy
        voltage:
          name: "Voltage L3"
          id: phase_c_volt
          filters: [ *throttle_avg, *pos ]
        phase_angle:
          name: "Phase Angle L3"
          filters: [ *throttle_avg, *pos ]

    ct_clamps:
      - phase_id: phase_a
        input: "A"  # Verify the CT going to this device input also matches the phase/leg
        power:
          id: phase_a_power
          accuracy_decimals: 2
          unit_of_measurement: "kW"
          filters: [ *throttle_avg_1s, *pos, *kW ]
      - phase_id: phase_b
        input: "B"  # Verify the CT going to this device input also matches the phase/leg
        power:
          id: phase_b_power
          accuracy_decimals: 2
          unit_of_measurement: "kW"
          filters: [ *throttle_avg_1s, *pos, *kW ]
      - phase_id: phase_c
        input: "C"  # Verify the CT going to this device input also matches the phase/leg
        power:
          id: phase_c_power
          accuracy_decimals: 2
          unit_of_measurement: "kW"
          filters: [ *throttle_avg_1s, *pos, *kW ]

      # Auto-detect phase by reading each CT against all three phases
      # Then template sensors below will select the most positive value
      # Internal sensors are throttled to reduce ESP32 load
      
      # Circuit 1 - read against all phases
      - { phase_id: phase_a, input:  "1", power: { id:  cir1_pa, filters: [ *kW, *throttle_internal ], internal: true } }
      - { phase_id: phase_b, input:  "1", power: { id:  cir1_pb, filters: [ *kW, *throttle_internal ], internal: true } }
      - { phase_id: phase_c, input:  "1", power: { id:  cir1_pc, filters: [ *kW, *throttle_internal ], internal: true } }
      
      # Circuit 2 - read against all phases
      - { phase_id: phase_a, input:  "2", power: { id:  cir2_pa, filters: [ *kW, *throttle_internal ], internal: true } }
      - { phase_id: phase_b, input:  "2", power: { id:  cir2_pb, filters: [ *kW, *throttle_internal ], internal: true } }
      - { phase_id: phase_c, input:  "2", power: { id:  cir2_pc, filters: [ *kW, *throttle_internal ], internal: true } }
      
      # Circuit 3 - read against all phases
      - { phase_id: phase_a, input:  "3", power: { id:  cir3_pa, filters: [ *kW, *throttle_internal ], internal: true } }
      - { phase_id: phase_b, input:  "3", power: { id:  cir3_pb, filters: [ *kW, *throttle_internal ], internal: true } }
      - { phase_id: phase_c, input:  "3", power: { id:  cir3_pc, filters: [ *kW, *throttle_internal ], internal: true } }
      
      # Circuit 4 - read against all phases
      - { phase_id: phase_a, input:  "4", power: { id:  cir4_pa, filters: [ *kW, *throttle_internal ], internal: true } }
      - { phase_id: phase_b, input:  "4", power: { id:  cir4_pb, filters: [ *kW, *throttle_internal ], internal: true } }
      - { phase_id: phase_c, input:  "4", power: { id:  cir4_pc, filters: [ *kW, *throttle_internal ], internal: true } }
      
      # Circuit 5 - read against all phases
      - { phase_id: phase_a, input:  "5", power: { id:  cir5_pa, filters: [ *kW, *throttle_internal ], internal: true } }
      - { phase_id: phase_b, input:  "5", power: { id:  cir5_pb, filters: [ *kW, *throttle_internal ], internal: true } }
      - { phase_id: phase_c, input:  "5", power: { id:  cir5_pc, filters: [ *kW, *throttle_internal ], internal: true } }
      
      # Circuit 6 - read against all phases
      - { phase_id: phase_a, input:  "6", power: { id:  cir6_pa, filters: [ *kW, *throttle_internal ], internal: true } }
      - { phase_id: phase_b, input:  "6", power: { id:  cir6_pb, filters: [ *kW, *throttle_internal ], internal: true } }
      - { phase_id: phase_c, input:  "6", power: { id:  cir6_pc, filters: [ *kW, *throttle_internal ], internal: true } }
      
      # Circuit 7 - read against all phases
      - { phase_id: phase_a, input:  "7", power: { id:  cir7_pa, filters: [ *kW, *throttle_internal ], internal: true } }
      - { phase_id: phase_b, input:  "7", power: { id:  cir7_pb, filters: [ *kW, *throttle_internal ], internal: true } }
      - { phase_id: phase_c, input:  "7", power: { id:  cir7_pc, filters: [ *kW, *throttle_internal ], internal: true } }
      
      # Circuit 8 - read against all phases
      - { phase_id: phase_a, input:  "8", power: { id:  cir8_pa, filters: [ *kW, *throttle_internal ], internal: true } }
      - { phase_id: phase_b, input:  "8", power: { id:  cir8_pb, filters: [ *kW, *throttle_internal ], internal: true } }
      - { phase_id: phase_c, input:  "8", power: { id:  cir8_pc, filters: [ *kW, *throttle_internal ], internal: true } }
      
      # Circuit 14 - read against all phases
      - { phase_id: phase_a, input: "14", power: { id: cir14_pa, filters: [ *kW, *throttle_internal ], internal: true } }
      - { phase_id: phase_b, input: "14", power: { id: cir14_pb, filters: [ *kW, *throttle_internal ], internal: true } }
      - { phase_id: phase_c, input: "14", power: { id: cir14_pc, filters: [ *kW, *throttle_internal ], internal: true } }
      
      # Circuit 15 - read against all phases
      - { phase_id: phase_a, input: "15", power: { id: cir15_pa, filters: [ *kW, *throttle_internal ], internal: true } }
      - { phase_id: phase_b, input: "15", power: { id: cir15_pb, filters: [ *kW, *throttle_internal ], internal: true } }
      - { phase_id: phase_c, input: "15", power: { id: cir15_pc, filters: [ *kW, *throttle_internal ], internal: true } }
      
      # Circuit 16 - read against all phases
      - { phase_id: phase_a, input: "16", power: { id: cir16_pa, filters: [ *kW, *throttle_internal ], internal: true } }
      - { phase_id: phase_b, input: "16", power: { id: cir16_pb, filters: [ *kW, *throttle_internal ], internal: true } }
      - { phase_id: phase_c, input: "16", power: { id: cir16_pc, filters: [ *kW, *throttle_internal ], internal: true } }
    on_update:
      then:
        - component.update: monitored_total_power
        - component.update: phase_a_cur
        - component.update: phase_b_cur
        - component.update: phase_c_cur
  - { platform: copy, name: "Power L1", source_id: phase_a_power, filters: *throttle_avg }
  - { platform: copy, name: "Power L2", source_id: phase_b_power, filters: *throttle_avg }
  - { platform: copy, name: "Power L3", source_id: phase_c_power, filters: *throttle_avg }
  - { platform: copy, name: "Current L1", source_id: phase_a_cur, accuracy_decimals: 2, filters: *throttle_avg }
  - { platform: copy, name: "Current L2", source_id: phase_b_cur, accuracy_decimals: 2, filters: *throttle_avg }
  - { platform: copy, name: "Current L3", source_id: phase_c_cur, accuracy_decimals: 2, filters: *throttle_avg }
  - { platform: copy, name: "Power Total", source_id: total_power, accuracy_decimals: 2, filters: *throttle_avg_1s }
  - { platform: copy, name: "Power Monitored Total", source_id: monitored_total_power, accuracy_decimals: 2, filters: *throttle_avg }
  # Auto-detected phase power sensors - select most positive value
  - platform: template
    name: "Power Circuit 1"
    id: cir1
    lambda: return max(max(id(cir1_pa).state, id(cir1_pb).state), id(cir1_pc).state);
    update_interval: 1s
    unit_of_measurement: "kW"
    device_class: power
    state_class: measurement
    accuracy_decimals: 2
    filters: [ *throttle_avg ]
  
  - platform: template
    name: "Power Circuit 2"
    id: cir2
    lambda: return max(max(id(cir2_pa).state, id(cir2_pb).state), id(cir2_pc).state);
    update_interval: 1s
    unit_of_measurement: "kW"
    device_class: power
    state_class: measurement
    accuracy_decimals: 2
    filters: [ *throttle_avg ]
  
  - platform: template
    name: "Power Circuit 3"
    id: cir3
    lambda: return max(max(id(cir3_pa).state, id(cir3_pb).state), id(cir3_pc).state);
    update_interval: 1s
    unit_of_measurement: "kW"
    device_class: power
    state_class: measurement
    accuracy_decimals: 2
    filters: [ *throttle_avg ]
  
  - platform: template
    name: "Power Circuit 4"
    id: cir4
    lambda: return max(max(id(cir4_pa).state, id(cir4_pb).state), id(cir4_pc).state);
    update_interval: 1s
    unit_of_measurement: "kW"
    device_class: power
    state_class: measurement
    accuracy_decimals: 2
    filters: [ *throttle_avg ]
  
  - platform: template
    name: "Power Circuit 5"
    id: cir5
    lambda: return max(max(id(cir5_pa).state, id(cir5_pb).state), id(cir5_pc).state);
    update_interval: 1s
    unit_of_measurement: "kW"
    device_class: power
    state_class: measurement
    accuracy_decimals: 2
    filters: [ *throttle_avg ]
  
  - platform: template
    name: "Power Circuit 6"
    id: cir6
    lambda: return max(max(id(cir6_pa).state, id(cir6_pb).state), id(cir6_pc).state);
    update_interval: 1s
    unit_of_measurement: "kW"
    device_class: power
    state_class: measurement
    accuracy_decimals: 2
    filters: [ *throttle_avg ]
  
  - platform: template
    name: "Power Circuit 7"
    id: cir7
    lambda: return max(max(id(cir7_pa).state, id(cir7_pb).state), id(cir7_pc).state);
    update_interval: 1s
    unit_of_measurement: "kW"
    device_class: power
    state_class: measurement
    accuracy_decimals: 2
    filters: [ *throttle_avg ]
  
  - platform: template
    name: "Power Circuit 8"
    id: cir8
    lambda: return max(max(id(cir8_pa).state, id(cir8_pb).state), id(cir8_pc).state);
    update_interval: 1s
    unit_of_measurement: "kW"
    device_class: power
    state_class: measurement
    accuracy_decimals: 2
    filters: [ *throttle_avg ]
  
  - platform: template
    name: "Power Circuit 14"
    id: cir14
    lambda: return max(max(id(cir14_pa).state, id(cir14_pb).state), id(cir14_pc).state);
    update_interval: 1s
    unit_of_measurement: "kW"
    device_class: power
    state_class: measurement
    accuracy_decimals: 2
    filters: [ *throttle_avg ]
  
  - platform: template
    name: "Power Circuit 15"
    id: cir15
    lambda: return max(max(id(cir15_pa).state, id(cir15_pb).state), id(cir15_pc).state);
    update_interval: 1s
    unit_of_measurement: "kW"
    device_class: power
    state_class: measurement
    accuracy_decimals: 2
    filters: [ *throttle_avg ]
  
  - platform: template
    name: "Power Circuit 16"
    id: cir16
    lambda: return max(max(id(cir16_pa).state, id(cir16_pb).state), id(cir16_pc).state);
    update_interval: 1s
    unit_of_measurement: "kW"
    device_class: power
    state_class: measurement
    accuracy_decimals: 2
    filters: [ *throttle_avg ]
  
  # Diagnostic sensors showing detected phase for each circuit (optional - for debugging)
  # Set internal: false to expose these sensors to Home Assistant
  - platform: template
    name: "Circuit 1 Detected Phase Value"
    internal: true  # Change to false to see in Home Assistant
    lambda: |-
      float pa = id(cir1_pa).state;
      float pb = id(cir1_pb).state;
      float pc = id(cir1_pc).state;
      if (pa >= pb && pa >= pc) return 1.0;
      if (pb >= pa && pb >= pc) return 2.0;
      return 3.0;
    update_interval: 10s
    accuracy_decimals: 0
    
  - platform: template
    name: "Circuit 2 Detected Phase Value"
    internal: true  # Change to false to see in Home Assistant
    lambda: |-
      float pa = id(cir2_pa).state;
      float pb = id(cir2_pb).state;
      float pc = id(cir2_pc).state;
      if (pa >= pb && pa >= pc) return 1.0;
      if (pb >= pa && pb >= pc) return 2.0;
      return 3.0;
    update_interval: 10s
    accuracy_decimals: 0
    
  - platform: template
    name: "Circuit 3 Detected Phase Value"
    internal: true  # Change to false to see in Home Assistant
    lambda: |-
      float pa = id(cir3_pa).state;
      float pb = id(cir3_pb).state;
      float pc = id(cir3_pc).state;
      if (pa >= pb && pa >= pc) return 1.0;
      if (pb >= pa && pb >= pc) return 2.0;
      return 3.0;
    update_interval: 10s
    accuracy_decimals: 0
    
  - platform: template
    name: "Circuit 4 Detected Phase Value"
    internal: true  # Change to false to see in Home Assistant
    lambda: |-
      float pa = id(cir4_pa).state;
      float pb = id(cir4_pb).state;
      float pc = id(cir4_pc).state;
      if (pa >= pb && pa >= pc) return 1.0;
      if (pb >= pa && pb >= pc) return 2.0;
      return 3.0;
    update_interval: 10s
    accuracy_decimals: 0
    
  - platform: template
    name: "Circuit 5 Detected Phase Value"
    internal: true  # Change to false to see in Home Assistant
    lambda: |-
      float pa = id(cir5_pa).state;
      float pb = id(cir5_pb).state;
      float pc = id(cir5_pc).state;
      if (pa >= pb && pa >= pc) return 1.0;
      if (pb >= pa && pb >= pc) return 2.0;
      return 3.0;
    update_interval: 10s
    accuracy_decimals: 0
    
  - platform: template
    name: "Circuit 6 Detected Phase Value"
    internal: true  # Change to false to see in Home Assistant
    lambda: |-
      float pa = id(cir6_pa).state;
      float pb = id(cir6_pb).state;
      float pc = id(cir6_pc).state;
      if (pa >= pb && pa >= pc) return 1.0;
      if (pb >= pa && pb >= pc) return 2.0;
      return 3.0;
    update_interval: 10s
    accuracy_decimals: 0
    
  - platform: template
    name: "Circuit 7 Detected Phase Value"
    internal: true  # Change to false to see in Home Assistant
    lambda: |-
      float pa = id(cir7_pa).state;
      float pb = id(cir7_pb).state;
      float pc = id(cir7_pc).state;
      if (pa >= pb && pa >= pc) return 1.0;
      if (pb >= pa && pb >= pc) return 2.0;
      return 3.0;
    update_interval: 10s
    accuracy_decimals: 0
    
  - platform: template
    name: "Circuit 8 Detected Phase Value"
    internal: true  # Change to false to see in Home Assistant
    lambda: |-
      float pa = id(cir8_pa).state;
      float pb = id(cir8_pb).state;
      float pc = id(cir8_pc).state;
      if (pa >= pb && pa >= pc) return 1.0;
      if (pb >= pa && pb >= pc) return 2.0;
      return 3.0;
    update_interval: 10s
    accuracy_decimals: 0
    
  - platform: template
    name: "Circuit 14 Detected Phase Value"
    internal: true  # Change to false to see in Home Assistant
    lambda: |-
      float pa = id(cir14_pa).state;
      float pb = id(cir14_pb).state;
      float pc = id(cir14_pc).state;
      if (pa >= pb && pa >= pc) return 1.0;
      if (pb >= pa && pb >= pc) return 2.0;
      return 3.0;
    update_interval: 10s
    accuracy_decimals: 0
    
  - platform: template
    name: "Circuit 15 Detected Phase Value"
    internal: true  # Change to false to see in Home Assistant
    lambda: |-
      float pa = id(cir15_pa).state;
      float pb = id(cir15_pb).state;
      float pc = id(cir15_pc).state;
      if (pa >= pb && pa >= pc) return 1.0;
      if (pb >= pa && pb >= pc) return 2.0;
      return 3.0;
    update_interval: 10s
    accuracy_decimals: 0
    
  - platform: template
    name: "Circuit 16 Detected Phase Value"
    internal: true  # Change to false to see in Home Assistant
    lambda: |-
      float pa = id(cir16_pa).state;
      float pb = id(cir16_pb).state;
      float pc = id(cir16_pc).state;
      if (pa >= pb && pa >= pc) return 1.0;
      if (pb >= pa && pb >= pc) return 2.0;
      return 3.0;
    update_interval: 10s
    accuracy_decimals: 0

  - platform: template
    lambda: return id(phase_a_power).state + id(phase_b_power).state + id(phase_c_power).state;
    update_interval: 1s
    id: total_power
    unit_of_measurement: "kW"
    device_class: power
    state_class: measurement
    accuracy_decimals: 2
  - platform: template
    lambda: !lambda |-
      return id( cir1).state +
        id( cir2).state +
        id( cir3).state +
        id( cir4).state +
        id( cir5).state +
        id( cir6).state +
        id( cir7).state +
        id( cir8).state +
        id(cir14).state +
        id(cir15).state +
        id(cir16).state;
    update_interval: never   # will be updated after all power sensors update via on_update trigger
    id: monitored_total_power
    device_class: power
    state_class: measurement
    unit_of_measurement: "kW"
    accuracy_decimals: 2
  - platform: total_daily_energy
    name: "Daily Energy Monitored Total"
    power_id: monitored_total_power
    accuracy_decimals: 2
    filters: *throttle_time
  - platform: total_daily_energy
    name: "Daily Energy Total"
    method: left
    state_class: total_increasing
    power_id: total_power
    accuracy_decimals: 2
    filters: *throttle_time
  - { power_id:  phase_a_power, platform: total_daily_energy, method: left, state_class: total_increasing, accuracy_decimals: 2, name:  "Daily Energy Phase L1", filters: *throttle_time }
  - { power_id:  phase_b_power, platform: total_daily_energy, method: left, state_class: total_increasing, accuracy_decimals: 2, name:  "Daily Energy Phase L2", filters: *throttle_time }
  - { power_id:  phase_c_power, platform: total_daily_energy, method: left, state_class: total_increasing, accuracy_decimals: 2, name:  "Daily Energy Phase L3", filters: *throttle_time }

  - { power_id:  cir1, platform: total_daily_energy, method: left, state_class: total_increasing, accuracy_decimals: 2, id: cir1_kwh, name:  "Daily Energy Circuit 1", filters: *throttle_time }
  - { power_id:  cir2, platform: total_daily_energy, method: left, state_class: total_increasing, accuracy_decimals: 2, id: cir2_kwh, name:  "Daily Energy Circuit 2", filters: *throttle_time }
  - { power_id: power_combined_1_2, platform: total_daily_energy, method: left, state_class: total_increasing, accuracy_decimals: 2, name: "Daily Energy Circuit 1+2", filters: *throttle_time }

  - { power_id:  cir3, platform: total_daily_energy, method: left, state_class: total_increasing, accuracy_decimals: 2, id: cir3_kwh, name:  "Daily Energy Circuit 3", filters: *throttle_time }
  - { power_id:  cir4, platform: total_daily_energy, method: left, state_class: total_increasing, accuracy_decimals: 2, id: cir4_kwh, name:  "Daily Energy Circuit 4", filters: *throttle_time }
  - { power_id: power_combined_3_4, platform: total_daily_energy, method: left, state_class: total_increasing, accuracy_decimals: 2, name: "Daily Energy Circuit 3+4", filters: *throttle_time }

  - { power_id:  cir5, platform: total_daily_energy, method: left, state_class: total_increasing, accuracy_decimals: 2, id: cir5_kwh, name:  "Daily Energy Circuit 5", filters: *throttle_time }

  - { power_id:  cir6, platform: total_daily_energy, method: left, state_class: total_increasing, accuracy_decimals: 2, id: cir6_kwh, name:  "Daily Energy Circuit 6", filters: *throttle_time }

  - { power_id:  cir7, platform: total_daily_energy, method: left, state_class: total_increasing, accuracy_decimals: 2, id: cir7_kwh, name:  "Daily Energy Circuit 7", filters: *throttle_time }

  - { power_id:  cir8, platform: total_daily_energy, method: left, state_class: total_increasing, accuracy_decimals: 2, id: cir8_kwh, name:  "Daily Energy Circuit 8", filters: *throttle_time }

#  - { power_id:  cir9, platform: total_daily_energy, accuracy_decimals: 2, id: cir9_kwh, name:  "Circuit 9 Daily Energy", filters: *throttle_time }
#  - { power_id: cir10, platform: total_daily_energy, accuracy_decimals: 2, id: cir10_kwh, name: "Circuit 10 Daily Energy", filters: *throttle_time }
#  - { power_id: cir11, platform: total_daily_energy, accuracy_decimals: 2, id: cir11_kwh, name: "Circuit 11 Daily Energy", filters: *throttle_time }
#  - { power_id: cir12, platform: total_daily_energy, accuracy_decimals: 2, id: cir12_kwh, name: "Circuit 12 Daily Energy", filters: *throttle_time }
#  - { power_id: cir13, platform: total_daily_energy, accuracy_decimals: 2, id: cir13_kwh, name: "Circuit 13 Daily Energy", filters: *throttle_time }

  - { power_id: cir14, platform: total_daily_energy, method: left, state_class: total_increasing, accuracy_decimals: 2, id: cir14_kwh, name: "Daily Energy Circuit 14", filters: *throttle_time }
  - { power_id: cir15, platform: total_daily_energy, method: left, state_class: total_increasing, accuracy_decimals: 2, id: cir15_kwh, name: "Daily Energy Circuit 15", filters: *throttle_time }
  - { power_id: power_combined_14_15, platform: total_daily_energy, method: left, state_class: total_increasing, accuracy_decimals: 2, name: "Daily Energy Circuit 14+15", filters: *throttle_time }

  - { power_id: cir16, platform: total_daily_energy, method: left, state_class: total_increasing, accuracy_decimals: 2, id: cir16_kwh, name: "Daily Energy Circuit 16", filters: *throttle_time }


  - platform: template
    lambda: return (id(phase_a_power).state * 1000) / id(phase_a_volt).state;
    id: phase_a_cur
    update_interval: never
    device_class: current
    state_class: measurement
    unit_of_measurement: "A"

  - platform: template
    lambda: return (id(phase_b_power).state * 1000) / id(phase_b_volt).state;
    id: phase_b_cur
    update_interval: never
    device_class: current
    state_class: measurement
    unit_of_measurement: "A"

  - platform: template
    lambda: return (id(phase_c_power).state * 1000) / id(phase_c_volt).state;
    id: phase_c_cur
    update_interval: never
    device_class: current
    state_class: measurement
    unit_of_measurement: "A"

  - platform: template
    name: "Power Circuit 1+2"
    id: power_combined_1_2
    lambda: return id(cir1).state + id(cir2).state;
    update_interval: 1s
    device_class: power
    state_class: measurement
    unit_of_measurement: "kW"
    accuracy_decimals: 2

  - platform: template
    name: "Power Circuit 3+4"
    lambda: return id(cir3).state + id(cir4).state;
    id: power_combined_3_4
    update_interval: 1s
    device_class: power
    state_class: measurement
    unit_of_measurement: "kW"
    accuracy_decimals: 2

  - platform: template
    name: "Power Circuit 14+15"
    id: power_combined_14_15
    lambda: return id(cir14).state + id(cir15).state;
    update_interval: 1s
    device_class: power
    state_class: measurement
    unit_of_measurement: "kW"
    accuracy_decimals: 2 

```

**Comment by gekkehenkie11** [comment]
LOL pretty nice to have see such good code by AI :) And it's pretty much the same as Emporia does in their stock FW. However keep in mind that in some specific cases this won't always work. If you don't have 180 degrees between the phases (so in a 3 phase environment) and have a device connected between two phases, then this won't work.

In a normal 180 degrees 2 phase system there won't be any problems (keep in mind that in some 2 phase systems there's actually 120 degrees between the phases, resulting in the same potential problem like I just described).

**Comment by gekkehenkie11** [comment]
BTW, Something I only just realized, to do this automatic detection, even in a regular 180 degrees phase setup: do make sure that your CT direction is correct!! Because you can't choose the highest absolute value, it has to be the highest positive value.

I actually only just realized because I was comparing my phase measurements in my breaker box with the values I had chosen in HA. And it turned out that one device, when turned on, would generate -1000 Watt, and about 340 Watt on the other phase. I hadnt even checked it and figured that -1000 watt would be the correct one (being the absolute highest), just had to multiply with -1 to correct for the CT wrong direction. But that was wrong! It turned out that the right value was actually that 340 Watt. 

So long story short, make sure your CT direction is correct when doing this!

**Reply by clowrey** [reply]
Thanks for noticing that! I have not really looked too closely as it's not my house this is at. 

It does work with 3 phase, that's the only system I have it on 120/208v 3p.

**Reply by gekkehenkie11** [reply]
Ha, nice, I'm on a similar system, 127/220V 3p. So also keep in mind what i said above, if you're connecting betwen 2 phases (in your case to get the 208V), then your code wont work.
